server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -683,6 +683,15 @@ private synchronized void consumeInternal(QuerySearchResult querySearchResult) {
             processedShards[querySearchResult.getShardIndex()] = querySearchResult.getSearchShardTarget();
         }
 
+        public TopDocs getBufferTopDocs() {
+            if (hasTopDocs) {
+                synchronized (this) {
+                    return topDocsBuffer[0];
```

Comment:
It's correct since any top hits that contains the requested size can be used but I agree that checking all top hits no matter what the buffer size could yield significant improvements. I was expecting to lower the batched_reduce_size by default in # so that this won't make a big difference but there are some discussions regarding the impact of setting this to a low value so making the change that you propose would make this decision less important for this optim. We also expect that this optimization will be really effective when the sort values are not randomly distributed so we could just check the worst hit of each shard and keep the minimum one which can be done linearly. I'll work on the change.

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/51852/commits/bf8ebde8430f12d9698f72e8d3ca54e61070993a